segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryCache.java

```diff
@@ -0,0 +1,439 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.pravega.common.Exceptions;
+import io.pravega.common.util.BufferView;
+import io.pravega.shared.protocol.netty.ByteBufWrapper;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.NonNull;
+import lombok.val;
+
+/**
+ * Represents a block-based {@link CacheStorage} that uses direct memory buffers to store data.
+ *
+ * Data Organization (actual values based on {@link CacheLayout.DefaultLayout}).
```

Comment:
A lot of this logic including breaking up memory into buffers and buffers into blocks of uniform size and managing them by reusing the same allocation is already implemented in Netty which this code is depending on anyway to allocate the actual buffers. `PooledByteBufAllocator` provides much of what this class implements, but has multiple pools for different size allocations, and has an optimized algorithm based on jemalloc to reclaim and reuse buffers. 

We should just use this class and pass in parameters to its constructor to get it to do what you want.

Reply:
I had a similar thought when I originally wrote this cache and I tried to use Netty's `PooledByteBufAllocator`. However I decided against it for the following reasons:
1. I would be tightly coupled with Netty. I don't want to depend on Netty and want to choose my own implementation at will. Currently I do use Netty's `ByteBuf` in the `DirectMemoryBuffer` class, but that could very easily be changed to Java's `ByteBuffer` (or whatever other implementation.
2. Netty has a bunch of logic to keep track of used and not-used blocks, but it does more than what I need. I wanted to keep the management of these buffers to a bare minimum to reduce the CPU needed to access them. In addition, I never deallocate the buffers and I want to implement my own policies as to how to place data (close to each other, etc).
3. I want to have the freedom to expand this into a disk-spillable cache. If I manage my own buffers it will make it easier for me to add such a functionality in the future. Tangling with Netty's allocator may actually make things harder.

Even without the three points above, the logic that deals with allocating the memory and managing it is so small that I did not find it worth it the effort to use Netty just for that. Most of the code in these 2 classes deal with organizing the data inside these buffers and providing access to them.

Reply:
I misunderstood how blocks worked. I didn't realize they were not shared across segments but were shared by appends within a segment.